New module: Gitlab webhook

drzraf opened the pull request:
##### SUMMARY
New module: **Gitlab web hooks aka Project hooks**

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
gitlab_webhook

##### ANSIBLE VERSION
```
ansible 2.5.2
```


##### ADDITIONAL INFORMATION
This module provides a way to setup Gitlab webhooks using Ansible.
Contrary to the first 3 gitlab modules (project, user and group), it relies upon the python-gitlab module instead of pyapi-gitlab (see #29310) 

Known bug: The Gitlab webhook "token" parameter is write-only (ie: not exposed on GET request).
That currently keeps module from being idempotent when a task set the `token` parameter.
See https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/7220#note_71782207

Commit:
New module: Gitlab webhook (#39991)

gundalow commented on lib/ansible/modules/source_control/gitlab_webhook.py:
```diff
@@ -0,0 +1,442 @@
+   - To forcefully remove the value of token for an already existing webhook, set it to the empty string.
+   - For convenience, webhook's URL may be used as the unique key for webhook edition/deletion instead of providing the hook_id.
+     The "solo" attribute can be used to handle cases where various hooks are defined for the same URL.
+     * state=absent,  none found           : error
```
This isn't valid yaml ` : `

gundalow commented on lib/ansible/modules/source_control/gitlab_webhook.py:
```diff
@@ -0,0 +1,442 @@
+   - If a token is specified, webhook is always modified since API does not provide a way to know current token
+   - To forcefully remove the value of token for an already existing webhook, set it to the empty string.
+   - For convenience, webhook's URL may be used as the unique key for webhook edition/deletion instead of providing the hook_id.
+     The "solo" attribute can be used to handle cases where various hooks are defined for the same URL.
```
I've move all of this into the EXAMPLES block

gundalow commented on lib/ansible/modules/source_control/gitlab_webhook.py:
```diff
@@ -0,0 +1,442 @@
+    validate_certs:
+        description:
+            - When using https if SSL certificate needs to be verified.
+        required: false
```
`required: false` is the default and can be removed

andreparames commented on lib/ansible/modules/source_control/gitlab_webhook.py:
```diff
@@ -0,0 +1,432 @@
+        deleted = []
+        for wh in remote_webhooks:
+            if wh.attributes['url'] == url:
+                deleted.append(project.hooks.delete(wh.attributes['id']))
```
I assume `project.hooks.delete` raises an Exception if it fails. Is the hook ID printed so the user can tell which one failed?

andreparames commented on lib/ansible/modules/source_control/gitlab_webhook.py:
```diff
@@ -0,0 +1,432 @@
+                h = HookHelper.update(project, local_webhook, remote_webhook[0])
+                diff['after'] = str(h)
+            except:
+                module.fail_json(changed=False, msg='Could not update webhook %s' % url)
```
Doesn't the API return any useful information about the error that could be passed to the user?

andreparames commented on lib/ansible/modules/source_control/gitlab_webhook.py:
```diff
@@ -0,0 +1,432 @@
+                diff = {'before': str(remote_webhook[0].attributes)}
+                h = HookHelper.update(project, local_webhook, remote_webhook[0])
+                diff['after'] = str(h)
+            except:
```
Bare `except:` is very rarely (if ever) what should be used. In this case, it means a `ctrl-c` will be caught and Ansible will keep going. I suggest using something more specific (at least `Exception`, which won't catch `ctrl-c`).

Commit:
handling Gitlab wrapper exceptions

drzraf commented:
@andreparames thank you for the review.
I tried to address the issue you mentionned by a new commit. Let me know if you want it to be squashed.

andreparames commented:
I'll leave the squashing decision to the Ansible core devs.

drzraf commented:
is there a specific protocol to get merged under 3 days? a ninja-merger somewhere? @maxamillion ?

dj-wasabi commented:
@drzraf I don't think there is one. I have closed several of my pr's a while ago after at least half year no response.

shipit

maxamillion commented:
It would be nice to have some unit tests for this, but I won't block on it.

shipit

gundalow commented on lib/ansible/modules/source_control/gitlab_webhook.py:
```diff
@@ -0,0 +1,452 @@
+---
+module: gitlab_webhook
+short_description: Creates, updates or deletes Gitlab webhook
+version_added: "2.6"
```
```suggestion
version_added: "2.8"
```

gundalow commented on lib/ansible/modules/source_control/gitlab_webhook.py:
```diff
@@ -0,0 +1,452 @@
+DOCUMENTATION = '''
+---
+module: gitlab_webhook
+short_description: Creates, updates or deletes Gitlab webhook
```
```suggestion
short_description: Creates, updates or deletes GitLab webhook
```

gundalow commented on lib/ansible/modules/source_control/gitlab_webhook.py:
```diff
@@ -0,0 +1,452 @@
+short_description: Creates, updates or deletes Gitlab webhook
+version_added: "2.6"
+description:
+  - When the webhook does not exists in this Gitlab project, it will be created.
```
```suggestion
  - When the webhook does not exists in this GitLab project, it will be created.
```

gundalow commented on lib/ansible/modules/source_control/gitlab_webhook.py:
```diff
@@ -0,0 +1,452 @@
+version_added: "2.6"
+description:
+  - When the webhook does not exists in this Gitlab project, it will be created.
+  - When the webhook does exists and state=absent, the webhook will be deleted.
```
```suggestion
  - When the webhook does exists and C(state=absent), the webhook will be deleted.
```

gundalow reviewed (commented):
`s/Gitlab/GitLab/g`

Commit:
misc

Shaps commented:
Isn't this what the `gitlab_hook` module does?

drzraf commented:
- This PR was provided May 11, 2018, have gone through the reviewing process and never got considered for merging.
- #40096 was provided May 14, 2018 and merged May 17, 2018

A couple of difference exist like:
- the fact that I tried hard to ensure idempotency, or at least allow for idempotency in case https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/7220#note_71782207 were to be fixed one day.
- the fact that this PR is based on a python library (which is itself tested) to delegate low-level work.

BTW I think the idea of testing against a live GitLab instance (#51490) is the right approach.

Lunik commented:
Compared to what I have done in #51141, the hook options are better handled in your module.

But now that `gitlab_hooks` is already part of Ansible, we can't remove it and replace it by your module. 

However, we can try to merge your modifications in #51141 ?

alxgu commented on lib/ansible/modules/source_control/gitlab_webhook.py:
```diff
@@ -0,0 +1,452 @@
+    except (gitlab.GitlabHttpError, gitlab.GitlabAuthenticationError, gitlab.GitlabGetError) as e:
+        module.fail_json(msg="Failed to connect to GitLab server", exception=to_native(e))
+
+    HookHelper = GitLabWebhook(module, git)
```
Why capitalized?
hookHelper or hook_helper

Lunik replied:
As long as the input and ouput of the module are in snail_case, I don't see any issue.

dagwieers commented:
We can still replace or remove the upstream **gitlab_hooks** module. It has not been shipped in an official release yet.

One objection to me is the fact that the module is plural, and we don't allow plural names.

dagwieers commented:
@Lunik Can we please get your big PR reviewed, tested and merged ASAP ? Because it doesn't help the open PRs if we are refactoring everything at once. IMO we should have first closed all existing PRs before attempting a refactor. Now we have this conflict of interest I'd rather get rid off.

Shaps commented:
I'd be happy to get the big PR merged but keep this in mind, once the big refactor is complete we can have a closer look at where this module is better and integrate later. We also don't have to necessarily remove the upstream module straightaway, we can alias it and follow the standard deprecation cycle

drzraf commented:
In terms of methodology I believe two statements are needed:
- whether preference goes to API-backed modules or not
- whether to enforce this preference for new modules

Then a refactor (and new PR reviews) would look like different  according to such a choice.

Lunik commented:
> In terms of methodology I believe two statements are needed:
> 
> * whether preference goes to API-backed modules or not
> * whether to enforce this preference for new modules
> 
> Then a refactor (and new PR reviews) would look like different according to such a choice.

In the [Community Wiki Guidelines](https://github.com/ansible/community/wiki/Gitlab-Module-Guidelines), we decided to use the `python-gitlab` library for future Gitlab modules. 
So your PR is following working group affinity.

Commit:
new param model

drzraf commented:
In reply to https://github.com/ansible/community/issues/431#issuecomment-461728960

As commented in the documentation the `solo` parameter exists to comply with least-surprises principle, w.r.t. logic **and** w.r.t. to GitLab API parameters.
Official position of upstream is that webhooks are NOT UNIQUE by their URL because two URL could theoretically be registered with two different tokens. See (and voice) here : https://gitlab.com/gitlab-org/gitlab-ce/issues/39864

About the somehow related idempotency topic, both this module and `gitlab_services` cope with https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/7220 . Best option is to voice/reopen/patch there. As soon as GitLab fixes this, both modules could be made  easily idempotent. I'll add this a ref' to upstream issue at https://github.com/ansible/community/wiki/Gitlab-Module-Guidelines

Commit:
sanity checks

Shaps commented on lib/ansible/modules/source_control/gitlab_webhook.py:
```diff
@@ -0,0 +1,442 @@
+            url=dict(required=False),
+            events=dict(required=False, type='list', elements='str', default=['push'],
+                        choices=['push', 'issues', 'confidential_issues', 'merge_requests', 'tag_push', 'note', 'job', 'pipeline', 'wiki_page']),
+            enable_ssl_verification=dict(required=False, default='yes', type='bool'),
```
I would rather call this parameter `hook_ssl_verify` or something along those lines since it's related to the hook ssl verification and it doesn't seem very clear to me ( I know it's mentioned in the docs ). Also I don't seem to be able to find where this is used?

Commit:
- changed param name from `enable_ssl_verification` to `hook_ssl_verify` and
  actually make use of it
- return value use `msg` for messages instead of `result`
- improve diff
- return the same message in check-mode as in real-mode
- improve initialization error message in case GitLab project does not exist

Shaps commented:
This looks good to me. @Lunik, any objections?

shipit

Lunik commented:
Wait a minute !
What about conflict with `gitlab_hook` module ?

Shaps commented:
@Lunik I thought since `gitlab_hook` had not officially shipped yet, we could have just merged this and aliased `gitlab_hook` to `gitlab_webhook`?

Lunik commented:
`gitlab_hook` is in Ansible since 2.6 so I don't understand what you mean ?

gundalow commented:
https://docs.ansible.com/ansible/devel/modules/gitlab_hook_module.html (singlular, not hook`s`) was added in Ansible 2.6, so should this be closed?

drzraf commented:
https://github.com/ansible/ansible/pull/39991#issuecomment-459173821

gundalow commented:
We now have a dedicated directory `/source_control/gitlab/` could the new module please be moved there.

Akasurde commented:
Hi @drzraf, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
